bridges/snowbridge/pallets/outbound-queue-v2/src/send_message_impl.rs

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+//! Implementation for [`snowbridge_core::outbound::SendMessage`]
+use super::*;
+use bridge_hub_common::AggregateMessageOrigin;
+use codec::Encode;
+use frame_support::{
+	ensure,
+	traits::{EnqueueMessage, Get},
+};
+use snowbridge_core::outbound::{
+	v2::{primary_governance_origin, Message, SendError, SendMessage},
+	SendMessageFeeProvider,
+};
+use sp_core::H256;
+use sp_runtime::BoundedVec;
+
+/// The maximal length of an enqueued message, as determined by the MessageQueue pallet
+pub type MaxEnqueuedMessageSizeOf<T> =
+	<<T as Config>::MessageQueue as EnqueueMessage<AggregateMessageOrigin>>::MaxMessageLen;
+
+impl<T> SendMessage for Pallet<T>
+where
+	T: Config,
+{
+	type Ticket = Message;
+
+	fn validate(
+		message: &Message,
+	) -> Result<(Self::Ticket, Fee<<Self as SendMessageFeeProvider>::Balance>), SendError> {
+		// The inner payload should not be too large
+		let payload = message.encode();
+		ensure!(
+			payload.len() < T::MaxMessagePayloadSize::get() as usize,
+			SendError::MessageTooLarge
+		);
+
+		let fee = Fee::from(Self::calculate_local_fee());
+
+		Ok((message.clone(), fee))
+	}
+
+	fn deliver(ticket: Self::Ticket) -> Result<H256, SendError> {
+		let origin = AggregateMessageOrigin::SnowbridgeV2(ticket.origin);
+
+		if ticket.origin != primary_governance_origin() {
+			ensure!(!Self::operating_mode().is_halted(), SendError::Halted);
+		}
+
+		let message =
+			BoundedVec::try_from(ticket.encode()).map_err(|_| SendError::MessageTooLarge)?;
+
+		T::MessageQueue::enqueue_message(message.as_bounded_slice(), origin);
+		Self::deposit_event(Event::MessageQueued { message: ticket.clone() });
```

Comment:
This isn't sufficient for dry-running, because the `message` at this stage in the processing pipeline isn't the same that is forwarded to the Ethereum gateway. Its missing information such as the gas limit. Without knowing the gas limits, we can't dry run properly on Ethereum.

In other words, we need the dry-run API to output the final message, the SCALE-encoded equivalent of this type: https://github.com/Snowfork/snowbridge/blob/fa9b8a7b3551cc9df8ee3239dc2830be5d84728e/contracts/src/v2/Types.sol#L8

This makes it much more easy for the UX to dry-run the resulting message on Ethereum.

We can ignore the message nonce for dry-run purposes on Ethereum.

So I you think you should explore adding a custom runtime API that takes the ExportMessage XCM, converts it to the final ethereum command.

Reply:
https://github.com/yrong/polkadot-sdk/pull/4/commits/0e27a22dd6a7af3130393841b0d67ca71fbf0035

Reply:
Meanwhile to remove the fee when [validate ExportMessage](https://github.com/yrong/polkadot-sdk/pull/4/commits/f9024d66a5e9e895ff93160998b19286bcb8193b) so that we can decrease the [delivery fee config](https://github.com/yrong/polkadot-sdk/blob/0e27a22dd6a7af3130393841b0d67ca71fbf0035/cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L694) on AH and make it as an user input.

More context in https://github.com/paritytech/polkadot-sdk/pull/4826#discussion_r1824022212